### DISCUSSION TITLE
FAQ

### DISCUSSION THREAD

**Discussion by ymmt2005** [ymmt2005/grpc-tutorial#2] [Q&A]
[FAQ.md](https://github.com/ymmt2005/grpc-tutorial/blob/main/FAQ.md) にまとめてあります。

**Comment by soharaki** [comment]
お忙しいところ恐縮ですが、チュートリアルを進めていて疑問に思ったので質問させてください。

 Computeは二つのRPC `Boot` と `Infer` を定義しています。  
どちらもクライアントから呼び出すときは 両方とも `context.Context` を必要としますが
サーバー側で `protoc` を使って作成したIFに従って実装するときは `Boot` は必要としません。

どうしてスタブの生成に差異があるのでしょうか。
疑問に思ったのでよろしくお願いいたします。

---

そもそもcontext自体、正直めっちゃ便利ー(棒)と雰囲気で使ってるのですが
gRPCで避けた方がいい使い方などあればご教授ください。

**Reply by ymmt2005** [reply]
そこをデザインしたわけではないので正確なことはわかりませんが、Boot は
server-side streaming を実装していて二番目の引数で `Context()` を呼べば
`context.Context` が入手できます。
https://pkg.go.dev/google.golang.org/grpc#ServerStream

おそらくは、stream 処理に必要なパラメーターを列挙するのを嫌って `grpc.ServerStream`
インタフェースにまとめる際、Context もまとめたというところではないでしょうか。

**Reply by soharaki** [reply]
元gRPCメンテナーの人も同じ趣旨のコメントを残してました!!
https://github.com/grpc/grpc-go/issues/151#issuecomment-90399581

streamとcontextが同じライフサイクルに位置していないと面倒くさそうというのはゆるゆるとわかり始めた次第です。
ex. https://github.com/grpc/grpc-go/issues/2159 

ご回答ありがとうございました。
